enterprise_access/apps/bffs/response_builder.py

```diff
@@ -0,0 +1,137 @@
+"""
+Response Builder Module for bffs app
+"""
+
+from enterprise_access.apps.bffs.serializers import LearnerDashboardResponseSerializer
+
+
+class BaseResponseBuilder:
+    """
+    A base response builder class that provides shared core functionality for different response builders.
+
+    The `BaseResponseBuilder` includes methods for building response data and can be extended by specific
+    response builders like `LearnerDashboardResponseBuilder` or `CourseResponseBuilder`.
+    """
+
+    def __init__(self, context):
+        """
+        Initializes the BaseResponseBuilder with a HandlerContext.
+
+        Args:
+            context (HandlerContext): The context object containing data, errors, and request information.
+        """
+        self.context = context
+
+    def build(self):
+        """
+        Builds the response data. This method should be overridden by subclasses to implement
+        specific response formatting logic.
+
+        Returns:
+            dict: A dictionary containing the response data.
+        """
+        raise NotImplementedError("Subclasses must implement the `build` method.")
+
+    def add_errors_warnings_to_response(self, response_data):
+        """
+        Adds any errors to the response data.
+        """
+        if self.context.errors:
+            response_data['errors'] = self.context.errors
+
+        if self.context.warnings:
+            response_data['warnings'] = self.context.warnings
+        return response_data
+
+    def get_status_code(self):
+        """
+        Gets the current status code from the context.
+
+        Returns:
+            int: The HTTP status code.
+        """
+        return self.context.status_code if hasattr(self.context, 'status_code') else 200
+
+
+class BaseLearnerResponseBuilder(BaseResponseBuilder):
+    """
+    A base response builder class for learner-focused routes.
+
+    The `BaseLearnerResponseBuilder` extends `BaseResponseBuilder` and provides shared core functionality
+    for building responses across all learner-focused page routes.
+    """
+
+    def common_response_logic(self, response_data):
+        """
+        Applies common response logic for learner-related responses.
+
+        Args:
+            response_data (dict): The initial response data.
+
+        Returns:
+            dict: The modified response data with common logic applied.
+        """
+        subscriptions_context = self.context.data.get('subscriptions', {})
+        enterprise_customer_user_subsidies = response_data.get('enterprise_customer_user_subsidies', {})
+        subscriptions = enterprise_customer_user_subsidies.get('subscriptions', {})
+        subscriptions.update(subscriptions_context)
+        enterprise_customer_user_subsidies.update({
+            'subscriptions': subscriptions,
+        })
+        response_data['enterprise_customer_user_subsidies'] = enterprise_customer_user_subsidies
```

Comment:
[consideration/thought question] Curious for your two cents here: I'm wondering if we should co-locate all subsidies-related metadata under `enterprise_customer_user_subsidies` within the `self.context.data`? Might co-locating subsidy types under a single dict in the handlers be worthwhile vs. having distinct properties (like `subscriptions`) at the top-level of `self.context.data`?

It _could_ make the handlers stuff more maintainable/easier to grok the context data if all subsidy types were grouped under `enterprise_customer_user_subsidies` before it gets to the response builder.

Reply:
Updated 👍🏽